// koa2-swagger-ui/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const handlebars_1 = __importDefault(require("handlebars"));
const lodash_1 = require("lodash");
const path_1 = __importDefault(require("path"));
const read_pkg_up_1 = __importDefault(require("read-pkg-up"));
const defaultOptions = {
    title: 'Swagger UI',
    oauthOptions: false,
    swaggerOptions: {
        // eslint-disable-next-line @typescript-eslint/camelcase
        dom_id: '#swagger-ui',
        url: 'https://petstore.swagger.io/v2/swagger.json',
        layout: 'StandaloneLayout',
    },
    routePrefix: '/docs',
    swaggerVersion: '',
    hideTopbar: false,
    favicon16: '/favicon-16x16.png',
    favicon32: '/favicon-32x32.png',
};
function koaSwagger(config = {}) {
    if (config.swaggerVersion === undefined) {
        const pkg = read_pkg_up_1.default.sync({ cwd: __dirname });
        if (pkg === undefined) {
            throw new Error('Package not found');
        }
        defaultOptions.swaggerVersion = pkg.packageJson.devDependencies['swagger-ui-dist'];
    }
    // Setup icons
    const extFavicon16 = config.favicon16;
    const extFavicon32 = config.favicon32;
    const favicon16Path = path_1.default.join(__dirname, defaultOptions.favicon16);
    const favicon32Path = path_1.default.join(__dirname, defaultOptions.favicon32);
    // Setup default options
    const options = lodash_1.defaultsDeep(config, defaultOptions);
    handlebars_1.default.registerHelper('json', context => JSON.stringify(context));
    handlebars_1.default.registerHelper('strfnc', fnc => fnc);
    handlebars_1.default.registerHelper('isset', function (conditional, opt) {
        return conditional ? opt.fn(this) : opt.inverse(this);
    });
    const index = handlebars_1.default.compile(fs_1.default.readFileSync(path_1.default.join(__dirname, './index.hbs'), 'utf-8'));
    // eslint-disable-next-line func-names
    return function koaSwaggerUi(ctx, next) {
        if (options.routePrefix === false || ctx.path === options.routePrefix) {
            ctx.type = 'text/html';
            ctx.body = index(options);
            return true;
        }
        if (extFavicon16 === undefined && ctx.path === defaultOptions.favicon16) {
            ctx.type = 'image/png';
            ctx.body = fs_1.default.createReadStream(favicon16Path);
            return true;
        }
        if (extFavicon32 === undefined && ctx.path === defaultOptions.favicon32) {
            ctx.type = 'image/png';
            ctx.body = fs_1.default.createReadStream(favicon32Path);
            return true;
        }
        return next();
    };
}
exports.default = koaSwagger;
module.exports = koaSwagger;
